tool/tsh/proxy.go

```diff
@@ -255,6 +258,91 @@ func mkLocalProxyCerts(certFile, keyFile string) ([]tls.Certificate, error) {
 	return []tls.Certificate{cert}, nil
 }
 
+func onProxyCommandAPP(cf *CLIConf) error {
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	appCerts, err := loadAppCertificate(tc, cf.AppName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	address, err := utils.ParseAddr(tc.WebProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	addr := "localhost:0"
+	if cf.LocalProxyPort != "" {
+		addr = fmt.Sprintf("127.0.0.1:%s", cf.LocalProxyPort)
+	}
+
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		Listener:           listener,
+		RemoteProxyAddr:    tc.WebProxyAddr,
+		Protocol:           alpncommon.ProtocolHTTP,
+		InsecureSkipVerify: cf.InsecureSkipVerify,
+		ParentContext:      cf.Context,
+		SNI:                address.Host(),
+		AWSCredentials:     nil,
+		Certs:              []tls.Certificate{appCerts},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	defer lp.Close()
+	go func() {
+		if err = lp.Start(cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local proxy.")
+		}
+	}()
+
+	fmt.Printf("Proxying connections to %s on %v\n", cf.AppName, lp.GetAddr())
+
+	<-cf.Context.Done()
+
+	return nil
```

Comment:
Any reason this is running `lp.Start` in a goroutine? It looks like [onProxyCommandDB](https://github.com/gravitational/teleport/blob/master/tool/tsh/proxy.go#L201) runs it inline. As it stands now if for some reason `lp.Start` fails it will log an error and the goroutine will exit but `tsh proxy app` will hang until the user manually kills the process.

Reply:
This is because the ALPN proxy wasn't correctly, handling context cancellations. I will fix this and add the commit.

Reply:
Should be fixed in https://github.com/gravitational/teleport/pull/10509/commits/f14d95c0614d4db1aaae694a95ccd403273be0a9. Originally I was having issues having ALPN Proxy close when context is canceled but following similar pattern to onProxyCommandDB it should work properly now.